### ISSUE TITLE
Extending custom layers functionality.

### ISSUE THREAD

**Issue by darionco** [uber/deck.gl#2204]
#### Background 
The `Component` class seals itself at the end of its constructor:
```
    ...
    // Seal the layer
    Object.seal(this);
  }
```
`Layer` extends `Component` and custom layers extend `Layer`. This architecture makes it hard to implement layers with any functionality that needs to keep track of variables:
```
export class CustomLayer extends Layer {
    constructor(props) {
        super(props);
        this.mRenderRefreshTime = 0; // <-- Throws error because object is sealed
    }
}
```
The problem is that sealing an object in the constructor makes it hard to truly extend an object and doesn't really protect the object from being extended, at the moment I am circumventing the limitation as follows:
```
class CustomLayerImp extends Layer {
    constructor(props) {
        super(props);
    }
    draw() {
        // ...
    }
    _preConstructor() {
        this.mRenderRefreshTime = 0;
    }
}
export const CustomLayer = function (props) {
    const layer = Object.create(CustomLayerImp.prototype);
    layer._preConstructor();
    CustomLayerImp.prototype.constructor.call(layer, props);
    return layer;
}
```
That is quite a hack just to add a member variable!

am I missing something?
is there a good reason to `Object.seal` the object in the constructor?
is there a better way of actually extending layers?
maybe add a `userData` member or similar to the `Component` class so developers can use it?

Any help/ideas will be appreciated.

Thanks in advance.

**Comment by ibgreen** [comment]
TL;DR - See if you can use the `layer.state` object

Full explanation:

An important thing to understand about layers is that they are disposable descriptors.

Every time you change your layer, you create a new layer and pass it to `Deck` or `DeckGL`.

Internally DeckGL matches new layers with old layers, and the old layers are released for garbage collection.

All permanent state is stored in the `layer.state` object, which "moves" between layers.

Does that make sense.

Marking this as a candidate for the upcoming FAQ.

**Comment by darionco** [comment]
Makes perfect sense, I misunderstood the [Layer Creation, Update and Destruction](https://github.com/uber/deck.gl/blob/master/docs/get-started/using-layers.md#layer-creation-update-and-destruction) section in [using-layers.md](https://github.com/uber/deck.gl/blob/master/docs/get-started/using-layers.md) and skipped [layer-lifecycle.md](https://github.com/uber/deck.gl/blob/master/docs/developer-guide/layer-lifecycle.md) where this is clearly explained.

I have a tendency to hack first and ask later, which led me to googling queries along the lines of `deck.gl "Object.seal"`, hopefully this post will help guide other rogue devs out there.

Thank you!
